### DISCUSSION TITLE
Cannot start td-agent service in docker container

### DISCUSSION THREAD

**Discussion by charlessensibill** [fluent/fluentd#3947] [Q&A] [answered]
### Describe the bug

I'm unable to run td-agent from within a node-16-bullseye-slim Docker container, but I think this will apply to any base image since I also tried with bullseye, buster-slim and buster

### To Reproduce

To reproduce the issue, start a container with this Dockerfile which tries to start td-agent and then use node to run start.js :

> start.js
```javascript
const nowTime = () => {
    console.log(new Date());
};

setInterval(() => {
    nowTime();
}, 1000);
```
> Dockerfile
```bash
FROM    node:16-bullseye-slim

RUN     apt-get update && apt-get -y install sudo
RUN     adduser --disabled-password chalu --uid 9001 --gecos '' --home /home/chalu --shell /bin/bash
RUN     adduser chalu sudo
RUN     echo '%chalu ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers

ARG     DEBIAN_FRONTEND=noninteractive
RUN     apt-get update \
        && apt-get install -y ca-certificates-java \
        && update-ca-certificates -f \
        && apt-get -qq update \
        && apt-get install -q --yes software-properties-common

RUN     apt-get -y install curl procps
RUN     curl -fsSL https://toolbelt.treasuredata.com/sh/install-debian-bullseye-td-agent4.sh | sh


ENV     APP_HOME /home/chalu/api
RUN     mkdir -p $APP_HOME

COPY    start.js /home/chalu/api/start.js

USER    root
CMD     ["sudo", "service", "td-agent", "start"]
```

### Expected behavior

I expected the `td-agent` service to start successfully, but instead I see `td-agent: unrecognized service` when I run the container from the built image

### Your Environment

```markdown
- Fluentd version:  N/A
- TD Agent version: v4 for Debian Bullseye. See https://docs.fluentd.org/installation/install-by-deb
- Operating system: I am building on Mac OS, but the container is to run on Debian 
- Kernel version: 21.6.0 (Mac)
```

### Your Configuration

```apache
<system>
  	# equal to -qq option
	log_level error
</system>

<source>
	@type syslog
	port 5140
	bind 0.0.0.0
	tag syslog
</source>

<match {syslog.**}>
	@type stdout
</match>
```
```


### Your Error Log

```shell
The service did not start, maybe that explains why the /var/log/td-agent/td-agent.log file did not even exist with `ls -ltr /var/log/td-agent/'
```

### Additional context

The goal was to run a simple Nodejs app in a container, such that logs from the app are tailed with td-agent and sent to a remote central server. 

Though to reproduce the issue, I used a simple Dockerfile with `CMD     ["sudo", "service", "td-agent", "start"]`, my application actually relies on a shell script that tries to 
1. start td-agent with `service td-agent start` just like it is able to start rsyslog with `service rsyslog start`.
2. start the Node js app

This shell script is then run in the Dockerfile as `CMD /path/to/script.sh`. 
Both `CMD /path/to/script.sh` and `CMD     ["sudo", "service", "td-agent", "start"]` have failed to get td-agent started.

I tried running `td-agent -q` instead of `service td-agent start` in the shell script but it starts and hijacks the flow such that the Nodejs app in start.js does not every get to run

**Comment by fujimotos** [comment, marked as answer]
My feeling is that what you're describing can be done by creating a bash script like this:

```bash
#!/bin/bash
td-agent -q -d 
node ./start.js
```

... and launch the container as follows:

```
ENTRYPOINT ["./init.sh"]
```

That said, I strongly recommend against setting up containers in this way.

> The goal was to run a simple Nodejs app in a container, such that logs from the app are tailed with td-agent and sent to a remote central server.

The whole point of Docker is to make it very easy to run services in separate
containers. If you want to run Fluentd, you really should just launch another
container for it.

The basic technical concept behind Docker is "one container, one app". I think
it kinda misses the point if we try to run both a Node app and Fluentd within the
same container.

> I expected the td-agent service to start successfully, but instead I see td-agent: unrecognized service when I run the container from the built image

I did not peek inside the `node:16-bullseye-slim` image, but most docker images
don't even include systemd. I guess this is the reason why it failed.

Please ask the image publisher for further details.

**Reply by charlessensibill** [reply]
Thanks for your response. 
I already tried the `td-agent -q -d` option and it seemed to get td-agent started (no errors displayed) but I don't know if it started it correctly / successfully since I don't see the `/var/log/td-agent/td-agent.log` file that should exist in the container with td-agent running.
 
Also, running `td-agent -q` again inside the container seems to start td-agent when I expected it to fail with  something like "already running" since the initial `td-agent -q -d` should have locked down resources like ports e.t.c. These seem to indicate that while using `td-agent -q -d` in the init script does not fail with any errors, it also does not seem to properly run td-agent as far as I can tell.

**Reply by fujimotos** [reply]
Try `td-agent --log td-agent.log -d td-agent.pid`; you need to give a file
path (to which Fluentd's PID is recorded) as an argument for `-d`.

If it still does not work, you should be able to find in td-agent.log.
